src/operator/grid_generator-inl.h

```diff
@@ -0,0 +1,314 @@
+
+namespace mxnet {
+namespace op {
+
+namespace grid {
+enum GridGeneratorOpInputs {kData};
+enum GridGeneratorOpOutputs {kOut, kGridDst, kTmp};
+enum GridGeneratorOpResource {kTempSpace};
+enum GridGeneratorTransformType {kAffine, kWarp};
+}
+
+struct GridGeneratorParam : public dmlc::Parameter<GridGeneratorParam> {
+  int transform_type;
+  TShape target_shape;
+  DMLC_DECLARE_PARAMETER(GridGeneratorParam) {
+    int shape[] = {0, 0};
+    DMLC_DECLARE_FIELD(transform_type)
+    .add_enum("affine", grid::kAffine)
+    .add_enum("warp", grid::kWarp)
+    .describe("transformation type\n    "
+              "if transformation type is affine, data is affine matrix : (batch, 6)\n    "
+              "if transformation type is warp, data is optical flow : (batch, 2, h, w)");
+    DMLC_DECLARE_FIELD(target_shape).set_default(TShape(shape, shape + 2))
+    .describe("if transformation type is affine, the operator need a target_shape : (H, W)\n    "
+              "if transofrmation type is warp, the operator will ignore target_shape");
+  }
+};
+
+template<typename xpu, typename DType>
+class GridGeneratorOp : public Operator {
+ public:
+  explicit GridGeneratorOp(GridGeneratorParam p) {
+    this->param_ = p;
+  }
+
+  virtual void Forward(const OpContext &ctx,
+                       const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1);
+    CHECK_EQ(out_data.size(), 3);
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    switch (param_.transform_type) {
+      case grid::kAffine: {
+        // if transform_type is affine, data is affine matrix, input shape : (batch, 2, 3)
+        Tensor<xpu, 4, DType> out = out_data[grid::kOut].get<xpu, 4, DType>(s);
+        Tensor<xpu, 2, DType> grid_dst = out_data[grid::kGridDst].get<xpu, 2, DType>(s);
+        Shape<3> data_shape = Shape3(out.size(0), 2, 3);
+        Tensor<xpu, 3, DType> data = in_data[grid::kData]
+          .get_with_shape<xpu, 3, DType>(data_shape, s);
+        Tensor<cpu, 2, DType> workspace =
+              ctx.requested[grid::kTempSpace].get_host_space_typed<2, DType>(
+              grid_dst.shape_);
+        Tensor<xpu, 3, DType> tmp = out_data[grid::kTmp].get<xpu, 3, DType>(s);
+        for (index_t i = 1; i <= workspace.size(1); i++) {
+          // grid dst coordinate is (x, y, 1)
+          workspace[0][i-1] = -1.0 + (i-1) % param_.target_shape[1] * 2.0 /
+                              (param_.target_shape[1] - 1);
+          workspace[1][i-1] = -1.0 + (i-1) / param_.target_shape[1] * 2.0 /
+                              (param_.target_shape[0] - 1);
+          workspace[2][i-1] = 1.0;
+        }
+        // create a sampling grid, tmp : (batch, 2, H x W)
+        Copy(grid_dst, workspace, grid_dst.stream_);

```

Comment:
Could we use GPU here? Like use `range` + element-wise operations to fill in the workspace